folio-spring-i18n/src/main/java/org/folio/spring/model/TranslationMap.java

```diff
@@ -0,0 +1,190 @@
+  }
+
+  /**
+   * Create a TranslationMap for a given locale based off of a file with a given fallback.
+   *
+   * @param locale the associated locale
+   * @param file the {@link TranslationFile TranslationFile} to read from
+   * @param fallback the TranslationMap to search if a given translation cannot be found
+   */
+  public TranslationMap(
+    Locale locale,
+    TranslationFile file,
+    @Nullable TranslationMap fallback
+  ) {
+    this(locale, file, file.getPatterns(), fallback);
+  }
+
+  /**
+   * Create a <em>default</em> TranslationMap for a given locale based off of a file.
+   * This constructor should only be used for default translation maps; all others
+   * should use the other constructor and provide a fallback.
+   *
+   * @param locale the associated locale
+   * @param file the {@link TranslationFile TranslationFile} to read from
+   */
+  public TranslationMap(Locale locale, TranslationFile file) {
+    this(locale, file, null);
+  }
+
+  /**
+   * Create a new TranslationMap for a given locale.  This is primarily used when cloning
+   * existing TranslationMaps for use as fallbacks for another (for caching reasons, this
+   * makes more sense than re-reading files each time)
+   *
+   * @param newLocale the locale for the new TranslationMap to use
+   * @return the TranslationMap with {@code newLocale}
+   */
+  public TranslationMap withLocale(Locale newLocale) {
+    return new TranslationMap(
+      newLocale,
+      this.file,
+      this.patterns,
+      this.fallback
+    );
+  }
+
+  /**
+   * Get the ICU format string associated with a given key for this translation.
+   *
+   * @param key the key to lookup
+   * @return one of the following, with this precedence:
+   *   <ol>
+   *     <li>the ICU format string for this locale, if it is known</li>
+   *     <li>the ICU format string for a locale with a worse {@code TranslationMatchQuality},
+   *       such as one without a specific country, recursing through the {@code fallback}</li>
+   *     <li>the ICU format string for the server's default locale</li>
+   *     <li>the provided key</li>
+   *   </ol>
+   */
+  public String get(String key) {
+    String pattern = this.patterns.get(key);
+
+    if (pattern == null && this.fallback != null) {
+      pattern = this.fallback.get(key);
+    }
+
+    if (pattern == null) {
+      log.error("Could not resolve key {} in any translation", key);
+      // fallback to translation key itself
+      pattern = key;
+    }
+
+    return pattern;
+  }
+
+  /**
+   * Format an ICU format string (found by its key), supplying a series of named arguments as key
+   * value pairs.  For example: {@code format("Hello {name}", "name", parameterValue)}
+   *
+   * @param key the key of the format string
+   * @param args pairs of keys and values to interpolate
+   * @return the formatted string
+   */
+  public String format(String key, Object... args) {
+    for (int i = 0; i < args.length; i++) {
```

Comment:
Small thing, but if nothing special is being done with `i`, a for-each loop (`for (Object arg : args) {`) is more idiomatic, since it's simpler and makes it harder for someone to come along and mess things up later (item 46 in _Effective Java_)

Reply:
Whoops, you're totally right! I missed that assignment